win/NvapiFansSvc/NvapiFansSvc.cpp

```
@@ -180,60 +183,128 @@ void LogSuccess(HANDLE event_log, std::wstring message) {
+        errmsg += "exception id: " + std::to_string(e.id) + "\n";
+        LogError(event_log, utf8_decode(errmsg));
+        return false;
+    }
+    service_config = draft_config;
+    return true;
 }
 
-bool loadConfig(HANDLE event_log, nlohmann::json& config) {
+bool loadConfig(const HANDLE &event_log, service_config_t &service_config) {
+
+    LPWSTR szPath;
 
-    LPWSTR szPath[MAX_PATH];
     // Get path for each computer, non-user specific and non-roaming data.
-    if (SHGetKnownFolderPath(FOLDERID_ProgramData, NULL, 0, szPath) >= 0) {
+    if (SHGetKnownFolderPath(FOLDERID_ProgramData, NULL, 0, &szPath) == S_OK) {
 
-        std::wstring appdata = *szPath;
+        std::wstring appdata = szPath;
         std::filesystem::path appdata_directory = appdata;
 
         std::filesystem::path config_directory_path = appdata_directory / CONFIG_DIR_NAME;
-
+ 
         if (!std::filesystem::exists(config_directory_path)) {
             LogError(event_log, L"Can't find config file parent: " + config_directory_path.wstring() + L". Creating it");
             bool res = std::filesystem::create_directory(config_directory_path);
             if (!res) {
                 LogError(event_log, L"Could not create directory: " + config_directory_path.wstring());
+                CoTaskMemFree(szPath);
                 return false;
             }
         }
         std::filesystem::path config_path = config_directory_path / CONFIG_FILE_NAME;
 
-
         if (!std::filesystem::exists(config_path)) {
-            LogError(event_log, L"Can't find config file: " + config_path.wstring() + L". Using defaults");
-            return false;
+            LogError(event_log, L"Can't find config file: " + config_path.wstring());
+        }
+        else {
+            LogInfo(event_log, L"I found: " + config_path.wstring());
+            const std::wstring config_path_w = config_path.wstring();
+            parseConfig(event_log, config_path_w, service_config);

```

Now that `parseConfig` won't modify `service_config` on failure, do you want to propagate the failure here as well?  Would a failure to load the configuration from disk would mean `loadConfig` failed?  If you're worried about not getting the "Applying config" message a few lines below, you could move that out of this function and into `serviceWorkerThread` to say what the current configuration is, regardless of whether a configuration load was successful or not.

(Remember, if you exit the function prematurely, you'll want to make sure you're still calling `CoTaskMemFree(szpath)`)

---

moved logging the config in the main loop of the worker thread